rust/worker/src/execution/operators/limit.rs

```diff
@@ -0,0 +1,317 @@
+pub struct LimitInput {
+    blockfile_provider: BlockfileProvider,
+    record_segment: Segment,
+    log_record: Chunk<LogRecord>,
+    log_oids: SignedRoaringBitmap,
+    compact_oids: SignedRoaringBitmap,
+    skip: u32,
+    fetch: Option<u32>,
+}
+
+impl LimitInput {
+    pub fn new(
+        blockfile_provider: BlockfileProvider,
+        record_segment: Segment,
+        log_record: Chunk<LogRecord>,
+        log_oids: SignedRoaringBitmap,
+        compact_oids: SignedRoaringBitmap,
+        skip: u32,
+        fetch: Option<u32>,
+    ) -> Self {
+        Self {
+            blockfile_provider,
+            record_segment,
+            log_record,
+            log_oids,
+            compact_oids,
+            skip,
+            fetch,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LimitOutput {
+    pub log_records: Chunk<LogRecord>,
+    pub offset_ids: RoaringBitmap,
+}
+
+#[derive(Error, Debug)]
+pub enum LimitError {
+    #[error("Error creating record segment reader {0}")]
+    RecordSegmentReaderCreationError(#[from] RecordSegmentReaderCreationError),
+    #[error("Error materializing logs {0}")]
+    LogMaterializationError(#[from] LogMaterializerError),
+    #[error("Error reading from record segment")]
+    RecordSegmentReaderError,
+}
+
+impl ChromaError for LimitError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            LimitError::RecordSegmentReaderCreationError(e) => e.code(),
+            LimitError::LogMaterializationError(e) => e.code(),
+            LimitError::RecordSegmentReaderError => ErrorCodes::Internal,
+        }
+    }
+}
+
+// This struct aims to help scanning a number of elements from a fixed offset in the imaginary
+// segment where the log is compacted and the element in the mask is ignored.
+struct SkipScanner<'me> {
+    log_oids: &'me RoaringBitmap,
+    record_segment: &'me RecordSegmentReader<'me>,
+    mask: &'me RoaringBitmap,
+}
+
+impl<'me> SkipScanner<'me> {
+    // Find the rank of the target offset id in the imaginary segment
+    async fn joint_rank(&self, target: u32) -> Result<usize, LimitError> {
+        // # of elements strictly less than target in materialized log
+        let log_rank =
+            self.log_oids.rank(target) as usize - self.log_oids.contains(target) as usize;
+        // # of elements strictly less than target in the record segment
+        let record_rank = self
+            .record_segment
+            .get_offset_id_rank(target)
+            .await
+            .map_err(|_| LimitError::RecordSegmentReaderError)?;
+        // # of elements strictly less than target in the mask
+        let mask_rank = self.mask.rank(target) as usize - self.mask.contains(target) as usize;
+        Ok(log_rank + record_rank - mask_rank)
+    }
+
+    // Skip to the start in log and record segment
+    // The following implementation is based on std implementation
+    async fn skip_to_start(&self, skip: usize) -> Result<(usize, usize), LimitError> {
```

Comment:
nit: I originally read skip as a verb, e.g. skip records until we reach start (rather than map some skip value to a start value)
maybe rename to something like
```suggestion
    async fn find_start(&self, skip: usize) -> Result<(usize, usize), LimitError> {
```
instead?

Reply:
Renamed in a later PR.